<<<<<<< HEAD
diff --git a/src/lib/node-modules-utils.js b/src/lib/node-modules-utils.js
<<<<<<< HEAD
index c1d6f1c1010020ce63c1736c528d2ecc461ca08d..2364cf74512d01f9b4b94679b678fd41309ce8c2 100644
--- a/src/lib/node-modules-utils.js
+++ b/src/lib/node-modules-utils.js
@@ -1,47 +1,23 @@
-import { createRequire } from 'module';
 import { checkResourceExists } from './resource-utils.js';
+import { resolveBareSpecifier, derivePackageRoot } from './walker-package-ranger.js';
 import fs from 'fs/promises';
 
-// TODO delete me and everything else in this file
-// https://github.com/ProjectEvergreen/greenwood/issues/684
-async function getNodeModulesLocationForPackage(packageName) {
-  let nodeModulesUrl;
-
-  // require.resolve may fail in the event a package has no main in its package.json
-  // so as a fallback, ask for node_modules paths and find its location manually
-  // https://github.com/ProjectEvergreen/greenwood/issues/557#issuecomment-923332104
-  // // https://stackoverflow.com/a/62499498/417806
-  const require = createRequire(import.meta.url);
-  const locations = require.resolve.paths(packageName);
-
-  for (const location in locations) {
-    const nodeModulesPackageRoot = `${locations[location]}/${packageName}`;
-    const packageJsonLocation = `${nodeModulesPackageRoot}/package.json`;
-
-    if (await checkResourceExists(new URL(`file://${packageJsonLocation}`))) {
-      nodeModulesUrl = nodeModulesPackageRoot;
-    }
+// take a "shortcut" pathname, e.g. /node_modules/lit/lit-html.js
+// and resolve it using import.meta.resolve
+function getResolvedHrefFromPathnameShortcut(pathname) {
+  const segments = pathname.replace('/node_modules/', '').split('/');
+  const hasScope = segments[0].startsWith('@');
+  const specifier = hasScope ? `${segments[0]}/${segments[1]}` : segments[0];
+  const resolved = resolveBareSpecifier(specifier);
+
+  if (resolved) {
+    const root = derivePackageRoot(resolved);
+
+    return `${root}${segments.slice(hasScope ? 2 : 1).join('/')}`;
+  } else {
+    // for example, local theme pack development
+    return `file://${pathname}`;
   }
-
-  if (!nodeModulesUrl) {
-    console.debug(`Unable to look up ${packageName} using NodeJS require.resolve.  Falling back to process.cwd()`);
-    nodeModulesUrl = new URL(`./node_modules/${packageName}`, `file://${process.cwd()}`).pathname;
-  }
-
-  return nodeModulesUrl;
-}
-
-// extract the package name from a URL like /node_modules/<some>/<package>/index.js
-function getPackageNameFromUrl(url) {
-  const packagePathPieces = url.split('node_modules/')[1].split('/'); // double split to handle node_modules within nested paths
-  let packageName = packagePathPieces.shift();
-
-  // handle scoped packages
-  if (packageName.indexOf('@') === 0) {
-    packageName = `${packageName}/${packagePathPieces.shift()}`;
-  }
-
-  return packageName;
 }
 
 async function getPackageJsonForProject({ userWorkspace, projectDirectory }) {
@@ -59,6 +35,5 @@ async function getPackageJsonForProject({ userWorkspace, projectDirectory }) {
 
 export {
   getPackageJsonForProject,
-  getNodeModulesLocationForPackage,
-  getPackageNameFromUrl
+  getResolvedHrefFromPathnameShortcut
 };
\ No newline at end of file
diff --git a/src/lib/resource-utils.js b/src/lib/resource-utils.js
index 6a78e464785b6dc11e909acf64f66db975e77909..0ab3458dc1542929a8b003509e6d63783b446967 100644
--- a/src/lib/resource-utils.js
+++ b/src/lib/resource-utils.js
@@ -1,5 +1,6 @@
 import fs from 'fs/promises';
 import { hashString } from './hashing-utils.js';
+import { getResolvedHrefFromPathnameShortcut } from './node-modules-utils.js';
 import htmlparser from 'node-html-parser';
 
 async function modelResource(context, type, src = undefined, contents = undefined, optimizationAttr = undefined, rawAttributes = undefined) {
@@ -7,9 +8,10 @@ async function modelResource(context, type, src = undefined, contents = undefine
   const extension = type === 'script' ? 'js' : 'css';
   let sourcePathURL;
 
+  getResolvedHrefFromPathnameShortcut
   if (src) {
     sourcePathURL = src.startsWith('/node_modules')
-      ? new URL(`.${src}`, projectDirectory)
+      ? new URL(getResolvedHrefFromPathnameShortcut(src))
       : src.startsWith('/')
         ? new URL(`.${src}`, userWorkspace)
         : new URL(`./${src.replace(/\.\.\//g, '').replace('./', '')}`, userWorkspace);
diff --git a/src/lib/walker-package-ranger.js b/src/lib/walker-package-ranger.js
index 3f5b0e870892bc59786ab466bcb7689fb5122539..0adb52cbb2f0f7aa8548855dbd0fdff7ac4e2255 100644
--- a/src/lib/walker-package-ranger.js
+++ b/src/lib/walker-package-ranger.js
@@ -3,11 +3,14 @@ import fs from 'fs';
 /* eslint-disable max-depth,complexity */
 // priority if from L -> R
 const SUPPORTED_EXPORT_CONDITIONS = ['import', 'module-sync', 'default'];
+const IMPORT_MAP_RESOLVED_PREFIX = '/~';
 const importMap = {};
 const diagnostics = {};
 
-function updateImportMap(key, value) {
-  importMap[key.replace('./', '')] = value.replace('./', '');
+function updateImportMap(key, value, resolvedRoot) {
+  if (!importMap[key.replace('./', '')]) {
+    importMap[key.replace('./', '')] = `${IMPORT_MAP_RESOLVED_PREFIX}${resolvedRoot.replace('file://', '')}${value.replace('./', '')}`;
+  }
 }
 
 // wrapper around import.meta.resolve to provide graceful error handling / logging
@@ -35,11 +38,27 @@ function resolveBareSpecifier(specifier) {
  *   root: 'file:///path/to/project/greenwood-lit-ssr/node_modules/.pnpm/lit-html@3.2.1/node_modules/lit-html/package.json'
  *  }
  */
-function derivePackageRoot(dependencyName, resolved) {
-  const root = resolved.slice(0, resolved.lastIndexOf(`/node_modules/${dependencyName}/`));
-  const derived = `${root}/node_modules/${dependencyName}/`;
+function derivePackageRoot(resolved) {
+  // can't rely on the specifier, for example in monorepos
+  // where @foo/bar may point to a non node_modules location
+  // e.g. packages/some-namespace/package.json
+  // so we walk backwards looking for nearest package.json
+  const segments = resolved
+    .replace('file://', '')
+    .split('/')
+    .filter(segment => segment !== '')
+    .reverse();
+  let root = resolved.replace(segments[0], '');
+
+  for (const segment of segments.slice(1)) {
+    if (fs.existsSync(new URL('./package.json', root).pathname)) {
+      break;
+    }
 
-  return derived;
+    root = root.replace(`${segment}/`, '');
+  }
+
+  return root;
 }
 
 // Helper function to convert export patterns to a regex (thanks ChatGPT :D)
@@ -109,7 +128,7 @@ async function walkExportPatterns(dependency, sub, subValue, resolvedRoot) {
         // https://unpkg.com/browse/@uswds/uswds@3.10.0/package.json
         const rootSubRelativePath = relativePath.replace(rootSubValueOffset, '');
 
-        updateImportMap(`${dependency}${rootSubOffset}${rootSubRelativePath}`, `/node_modules/${dependency}${relativePath}`);
+        updateImportMap(`${dependency}${rootSubOffset}${rootSubRelativePath}`, relativePath, resolvedRoot);
       }
     });
   }
@@ -117,18 +136,18 @@ async function walkExportPatterns(dependency, sub, subValue, resolvedRoot) {
   walkDirectoryForExportPatterns(new URL(`.${rootSubValueOffset}/`, resolvedRoot));
 }
 
-function trackExportConditions(dependency, exports, sub, condition) {
+function trackExportConditions(dependency, exports, sub, condition, resolvedRoot) {
   if (typeof exports[sub] === 'object') {
     // also check for nested conditions of conditions, default to default for now
     // https://unpkg.com/browse/@floating-ui/dom@1.6.12/package.json
     if (sub === '.') {
-      updateImportMap(dependency, `/node_modules/${dependency}/${exports[sub][condition].default ?? exports[sub][condition]}`);
+      updateImportMap(dependency, `${exports[sub][condition].default ?? exports[sub][condition]}`, resolvedRoot);
     } else {
-      updateImportMap(`${dependency}/${sub}`, `/node_modules/${dependency}/${exports[sub][condition].default ?? exports[sub][condition]}`);
+      updateImportMap(`${dependency}/${sub}`, `${exports[sub][condition].default ?? exports[sub][condition]}`, resolvedRoot);
     }
   } else {
     // https://unpkg.com/browse/redux@5.0.1/package.json
-    updateImportMap(dependency, `/node_modules/${dependency}/${exports[sub][condition]}`);
+    updateImportMap(dependency, `${exports[sub][condition]}`);
   }
 }
 
@@ -151,7 +170,7 @@ async function walkPackageForExports(dependency, packageJson, resolvedRoot) {
         for (const condition of SUPPORTED_EXPORT_CONDITIONS) {
           if (exports[sub][condition]) {
             matched = true;
-            trackExportConditions(dependency, exports, sub, condition);
+            trackExportConditions(dependency, exports, sub, condition, resolvedRoot);
             break;
           }
         }
@@ -163,16 +182,16 @@ async function walkPackageForExports(dependency, packageJson, resolvedRoot) {
       } else {
         // handle (unconditional) subpath exports
         if (sub === '.') {
-          updateImportMap(dependency, `/node_modules/${dependency}/${exports[sub]}`);
+          updateImportMap(dependency, `${exports[sub]}`, resolvedRoot);
         } else if (sub.indexOf('*') >= 0) {
           await walkExportPatterns(dependency, sub, exports[sub], resolvedRoot);
         } else {
-          updateImportMap(`${dependency}/${sub}`, `/node_modules/${dependency}/${exports[sub]}`);
+          updateImportMap(`${dependency}/${sub}`, `${exports[sub]}`, resolvedRoot);
         }
       }
     }
   } else if (module || main) {
-    updateImportMap(dependency, `/node_modules/${dependency}/${module ?? main}`);
+    updateImportMap(dependency, `${module ?? main}`, resolvedRoot);
   } else {
     // ex: https://unpkg.com/browse/uuid@3.4.0/package.json
     diagnostics[dependency] = `WARNING: No supported entry point detected for => \`${dependency}\``;
@@ -186,7 +205,7 @@ async function walkPackageJson(packageJson = {}) {
       const resolved = resolveBareSpecifier(dependency);
 
       if (resolved) {
-        const resolvedRoot = derivePackageRoot(dependency, resolved);
+        const resolvedRoot = derivePackageRoot(resolved);
         const resolvedPackageJson = (await import(new URL('./package.json', resolvedRoot), { with: { type: 'json' } })).default;
 
         walkPackageForExports(dependency, resolvedPackageJson, resolvedRoot);
@@ -196,7 +215,7 @@ async function walkPackageJson(packageJson = {}) {
             const resolved = resolveBareSpecifier(dependency);
 
             if (resolved) {
-              const resolvedRoot = derivePackageRoot(dependency, resolved);
+              const resolvedRoot = derivePackageRoot(resolved);
               const resolvedPackageJson = (await import(new URL('./package.json', resolvedRoot), { with: { type: 'json' } })).default;
 
               walkPackageForExports(dependency, resolvedPackageJson, resolvedRoot);
@@ -246,5 +265,8 @@ function mergeImportMap(html = '', map = {}, shouldShim = false) {
 
 export {
   walkPackageJson,
-  mergeImportMap
+  mergeImportMap,
+  resolveBareSpecifier,
+  derivePackageRoot,
+  IMPORT_MAP_RESOLVED_PREFIX
 };
\ No newline at end of file
diff --git a/src/plugins/resource/plugin-node-modules.js b/src/plugins/resource/plugin-node-modules.js
index a80b54d6e094762056595635a7c9d95eab139260..92125cdcdaac5584c7a31b92ad3814b9529412b6 100644
--- a/src/plugins/resource/plugin-node-modules.js
+++ b/src/plugins/resource/plugin-node-modules.js
@@ -6,11 +6,10 @@
 import { checkResourceExists } from '../../lib/resource-utils.js';
 import fs from 'fs/promises';
 import { nodeResolve } from '@rollup/plugin-node-resolve';
-import { getNodeModulesLocationForPackage, getPackageJsonForProject, getPackageNameFromUrl } from '../../lib/node-modules-utils.js';
-import { resolveForRelativeUrl } from '../../lib/resource-utils.js';
+import { getPackageJsonForProject, getResolvedHrefFromPathnameShortcut } from '../../lib/node-modules-utils.js';
 import { ResourceInterface } from '../../lib/resource-interface.js';
 import { mergeImportMap } from '../../lib/walker-package-ranger.js';
-import { walkPackageJson } from '../../lib/walker-package-ranger.js';
+import { walkPackageJson, IMPORT_MAP_RESOLVED_PREFIX } from '../../lib/walker-package-ranger.js';
 
 let generatedImportMap;
 
@@ -22,42 +21,36 @@ class NodeModulesResource extends ResourceInterface {
   }
 
   async shouldResolve(url) {
-    return url.pathname.indexOf('/node_modules/') === 0;
+    const { pathname } = url;
+
+    return pathname.startsWith(IMPORT_MAP_RESOLVED_PREFIX) || pathname.startsWith('/node_modules/');
   }
 
   async resolve(url) {
     const { projectDirectory } = this.compilation.context;
     const { pathname, searchParams } = url;
-    const packageName = getPackageNameFromUrl(pathname);
-    const absoluteNodeModulesLocation = await getNodeModulesLocationForPackage(packageName);
-    const packagePathPieces = pathname.split('node_modules/')[1].split('/'); // double split to handle node_modules within nested paths
-    // use node modules resolution logic first, else hope for the best from the root of the project
-    const absoluteNodeModulesPathname = absoluteNodeModulesLocation
-      ? `${absoluteNodeModulesLocation}${packagePathPieces.join('/').replace(packageName, '')}`
-      : (await resolveForRelativeUrl(url, projectDirectory)).pathname;
+    const fromImportMap = pathname.startsWith(IMPORT_MAP_RESOLVED_PREFIX);
+    const isNodeModulesPathnameShortcut = pathname.startsWith('/node_modules/');
+    const resolvedHref = fromImportMap
+      ? pathname.replace(IMPORT_MAP_RESOLVED_PREFIX, 'file://')
+      : isNodeModulesPathnameShortcut
+        ? getResolvedHrefFromPathnameShortcut(pathname)
+        : new URL(`.${pathname}`, projectDirectory).href; // worst case fall back, assume project root
     const params = searchParams.size > 0
       ? `?${searchParams.toString()}`
       : '';
 
-    return new Request(`file://${absoluteNodeModulesPathname}${params}`);
+    return new Request(`${resolvedHref}${params}`);
   }
 
   async shouldServe(url) {
-    const { href, pathname, protocol } = url;
-    const extension = pathname.split('.').pop();
-    const existsAsJs = protocol === 'file:' && await checkResourceExists(new URL(`${href}.js`));
+    const { href, protocol } = url;
 
-    return extension === 'mjs'
-      || extension === '' && existsAsJs
-      || extension === 'js' && url.pathname.startsWith('/node_modules/');
+    return protocol === 'file:' && await checkResourceExists(new URL(href));
   }
 
   async serve(url) {
-    const pathname = url.pathname;
-    const urlExtended = pathname.split('.').pop() === ''
-      ? new URL(`file://${pathname}.js`)
-      : url;
-    const body = await fs.readFile(urlExtended, 'utf-8');
+    const body = await fs.readFile(url, 'utf-8');
 
     return new Response(body, {
       headers: new Headers({
=======
index 05226f66bf42fc0a07588d373b14932cbeb9ff7f..a86924565e964815769e1fc788bf5c6cb6454083 100644
--- a/src/lib/node-modules-utils.js
+++ b/src/lib/node-modules-utils.js
@@ -29,6 +29,28 @@ async function getNodeModulesLocationForPackage(packageName) {
   return nodeModulesUrl;
 }
 
+function resolveBareSpecifier(specifier) {
+  let resolvedPath;
+
+  // sometimes a package.json has no main field :/
+  // https://unpkg.com/browse/@types/trusted-types@2.0.7/package.json
+  try {
+    resolvedPath = import.meta.resolve(specifier);
+  } catch (e) {
+    // TODO console.log(`WARNING: unable to resolve specifier \`${specifier}\``);
+  }
+
+  return resolvedPath;
+}
+
+function resolveRootForSpecifier(specifier) {
+  const resolved = resolveBareSpecifier(specifier);
+
+  if (resolved) {
+    return `${resolved.split(specifier)[0]}${specifier}/`;
+  }
+}
+
 // extract the package name from a URL like /node_modules/<some>/<package>/index.js
 function getPackageNameFromUrl(url) {
   const packagePathPieces = url.split('node_modules/')[1].split('/'); // double split to handle node_modules within nested paths
@@ -58,5 +80,7 @@ async function getPackageJson({ userWorkspace, projectDirectory }) {
 export {
   getNodeModulesLocationForPackage,
   getPackageJson,
-  getPackageNameFromUrl
+  getPackageNameFromUrl,
+  resolveBareSpecifier,
+  resolveRootForSpecifier
 };
\ No newline at end of file
=======
>>>>>>> a7bb2ab (latest upstream patches)
diff --git a/src/lib/walker-package-ranger.js b/src/lib/walker-package-ranger.js
index a92bb3d29496a453f1714e5cd32d1800c94844be..ded1034c929edc462ab547dc9b16ca8cf26cf5a7 100644
--- a/src/lib/walker-package-ranger.js
+++ b/src/lib/walker-package-ranger.js
@@ -1,220 +1,216 @@
-/* eslint-disable max-depth,complexity */
-import * as acorn from 'acorn';
 import fs from 'fs';
-import { getNodeModulesLocationForPackage } from './node-modules-utils.js';
-import path from 'path';
-import * as walk from 'acorn-walk';
 
+/* eslint-disable max-depth,complexity */
+// priority if from L -> R
+const SUPPORTED_EXPORT_CONDITIONS = ['import', 'module-sync', 'default'];
 const importMap = {};
 
-const updateImportMap = (entry, entryPath) => {
+function updateImportMap(key, value) {
+  importMap[key.replace('./', '')] = value.replace('./', '');
+}
 
-  if (path.extname(entryPath) === '') {
-    entryPath = `${entryPath}.js`;
+// wrapper around import.meta.resolve to provide graceful error handling / logging
+// as sometimes a package.json has no main field :/
+// https://unpkg.com/browse/@types/trusted-types@2.0.7/package.json
+// https://github.com/nodejs/node/issues/49445#issuecomment-2484334036
+function resolveBareSpecifier(specifier) {
+  let resolvedPath;
+
+  try {
+    resolvedPath = import.meta.resolve(specifier);
+  } catch (e) {
+    // console.log({ e });
+    // TODO console.log(`WARNING: unable to resolve specifier \`${specifier}\``);
   }
 
-  // handle WIn v Unix-style path separators and force to /
-  importMap[entry.replace(/\\/g, '/')] = entryPath.replace(/\\/g, '/');
-};
+  return resolvedPath;
+}
 
-// handle ESM paths that have varying levels of nesting, e.g. export * from '../../something.js'
-// https://github.com/ProjectEvergreen/greenwood/issues/820
-async function resolveRelativeSpecifier(specifier, modulePath, dependency) {
-  const absoluteNodeModulesLocation = await getNodeModulesLocationForPackage(dependency);
+/*
+ * Find root directory for a package based on result of import.meta.resolve, since dependencyName could show in multiple places
+ * until this becomes a thing - https://github.com/nodejs/node/issues/49445
+ * {
+ *   dependencyName: 'lit-html',
+ *   resolved: 'file:///path/to/project/greenwood-lit-ssr/node_modules/.pnpm/lit-html@3.2.1/node_modules/lit-html/node/lit-html.js',
+ *   root: 'file:///path/to/project/greenwood-lit-ssr/node_modules/.pnpm/lit-html@3.2.1/node_modules/lit-html/package.json'
+ *  }
+ */
+function derivePackageRoot(dependencyName, resolved) {
+  const root = resolved.slice(0, resolved.lastIndexOf(`/node_modules/${dependencyName}/`));
+  const derived = `${root}/node_modules/${dependencyName}/`;
+
+  return derived;
+}
+
+// Helper function to convert export patterns to a regex (thanks ChatGPT :D)
+function globToRegex(pattern) {
+  // Escape special regex characters
+  pattern = pattern.replace(/[.+^${}()|[\]\\]/g, '\\$&');
+
+  // Replace glob `*` with regex `[^/]*` (any characters except slashes)
+  pattern = pattern.replace(/\*/g, '[^/]*');
 
-  // handle WIn v Unix-style path separators and force to /
-  return `${dependency}${path.join(path.dirname(modulePath), specifier).replace(/\\/g, '/').replace(absoluteNodeModulesLocation.replace(/\\/g, '/', ''), '')}`;
+  // Replace glob `**` with regex `(.*)` (zero or more directories or files)
+  // pattern = pattern.replace(/\*\*/g, '(.*)');
+
+  // Return the final regex
+  return new RegExp('^' + pattern + '$');
 }
 
-async function getPackageEntryPath(packageJson) {
-  let entry = packageJson.exports
-    ? Object.keys(packageJson.exports) // first favor export maps first
-    : packageJson.module // next favor ESM entry points
-      ? packageJson.module
-      : packageJson.main && packageJson.main !== '' // then favor main
-        ? packageJson.main
-        : 'index.js'; // lastly, fallback to index.js
-
-  // use .mjs version if it exists, for packages like redux
-  if (!Array.isArray(entry) && fs.existsSync(`${await getNodeModulesLocationForPackage(packageJson.name)}/${entry.replace('.js', '.mjs')}`)) {
-    entry = entry.replace('.js', '.mjs');
+// convert path to its lowest common root
+// e.g. ./img/path/*/index.js -> /img/path
+// https://unpkg.com/browse/@uswds/uswds@3.10.0/package.json
+function patternRoot(pattern) {
+  const segments = pattern.split('/').filter((segment) => segment !== '.');
+  let root = '';
+
+  for (const segment of segments) {
+    // is there a better way to fuzzy test for a filename other than checking for a dot?
+    if (segment.indexOf('*') < 0 && segment.indexOf('.') < 0) {
+      root += `/${segment}`;
+    } else {
+      break;
+    }
   }
 
-  return entry;
+  return root;
 }
 
-async function walkModule(modulePath, dependency) {
-  const moduleContents = fs.readFileSync(modulePath, 'utf-8');
-
-  walk.simple(acorn.parse(moduleContents, {
-    ecmaVersion: '2020',
-    sourceType: 'module'
-  }), {
-    async ImportDeclaration(node) {
-      let { value: sourceValue } = node.source;
-      const absoluteNodeModulesLocation = await getNodeModulesLocationForPackage(dependency);
-      const isBarePath = sourceValue.indexOf('http') !== 0 && sourceValue.charAt(0) !== '.' && sourceValue.charAt(0) !== path.sep;
-      const hasExtension = path.extname(sourceValue) !== '';
-
-      if (isBarePath && !hasExtension) {
-        if (!importMap[sourceValue]) {
-          updateImportMap(sourceValue, `/node_modules/${sourceValue}`);
-        }
+/*
+ * https://nodejs.org/api/packages.html#subpath-patterns
+ *
+ * Examples
+ * "./icons/*": "./icons/*" - https://unpkg.com/browse/@spectrum-web-components/icons-workflow@1.0.1/package.json
+ * "./components/*": "./dist/components/*.js" - https://unpkg.com/browse/@uswds/web-components@0.0.1-alpha/package.json
+ * "./src/components/*": "./src/components/* /index.js - https://unpkg.com/browse/@uswds/web-components@0.0.1-alpha/package.json
+ */
+async function walkExportPatterns(dependency, sub, subValue, resolvedRoot) {
+  // find the "deepest" segment we can start from
+  // to avoid unnecessary file scanning / crawling
+  const rootSubValueOffset = patternRoot(subValue);
+
+  // ideally we can use fs.glob when it comes out of experimental
+  // https://nodejs.org/docs/latest-v22.x/api/fs.html#fspromisesglobpattern-options
+  function walkDirectoryForExportPatterns(directoryUrl) {
+    const filesInDir = fs.readdirSync(directoryUrl);
+
+    filesInDir.forEach(file => {
+      const filePathUrl = new URL(`./${file}`, directoryUrl);
+      const stat = fs.statSync(filePathUrl);
+      const pattern = `${resolvedRoot}${subValue.replace('./', '')}`;
+      const regexPattern = globToRegex(pattern);
+
+      if (stat.isDirectory()) {
+        walkDirectoryForExportPatterns(new URL(`./${file}/`, directoryUrl));
+      } else if (regexPattern.test(filePathUrl.href)) {
+        const rootSubOffset = patternRoot(sub);
+        const relativePath = filePathUrl.href.replace(resolvedRoot, '');
+
+        updateImportMap(`${dependency}${rootSubOffset}/${file}`, `/node_modules/${dependency}/${relativePath}`);
+      }
+    });
+  }
 
-        await walkPackageJson(path.join(absoluteNodeModulesLocation, 'package.json'));
-      } else if (isBarePath) {
-        updateImportMap(sourceValue, `/node_modules/${sourceValue}`);
-      } else {
-        // walk this module for all its dependencies
-        sourceValue = !hasExtension
-          ? `${sourceValue}.js`
-          : sourceValue;
+  walkDirectoryForExportPatterns(new URL(`.${rootSubValueOffset}/`, resolvedRoot));
+}
 
-        if (fs.existsSync(path.join(absoluteNodeModulesLocation, sourceValue))) {
-          const entry = `/node_modules/${await resolveRelativeSpecifier(sourceValue, modulePath, dependency)}`;
-          await walkModule(path.join(absoluteNodeModulesLocation, sourceValue), dependency);
+function trackExportConditions(dependency, exports, sub, condition) {
+  if (typeof exports[sub] === 'object') {
+    // also check for nested conditions of conditions, default to default for now
+    // https://unpkg.com/browse/@floating-ui/dom@1.6.12/package.json
+    if (sub === '.') {
+      updateImportMap(dependency, `/node_modules/${dependency}/${exports[sub][condition].default ?? exports[sub][condition]}`);
+    } else {
+      updateImportMap(`${dependency}/${sub}`, `/node_modules/${dependency}/${exports[sub][condition].default ?? exports[sub][condition]}`);
+    }
+  } else {
+    // https://unpkg.com/browse/redux@5.0.1/package.json
+    updateImportMap(dependency, `/node_modules/${dependency}/${exports[sub][condition]}`);
+  }
+}
 
-          updateImportMap(path.join(dependency, sourceValue), entry);
+// https://nodejs.org/api/packages.html#conditional-exports
+async function walkPackageForExports(dependency, packageJson, resolvedRoot) {
+  const { exports, module, main } = packageJson;
+
+  // favor exports over main / module
+  if (exports) {
+    for (const sub in exports) {
+      /*
+       * test for conditional subpath exports
+       * 1. import
+       * 2. module-sync
+       * 3. default
+       */
+      if (typeof exports[sub] === 'object') {
+        let matched = false;
+
+        for (const condition of SUPPORTED_EXPORT_CONDITIONS) {
+          if (exports[sub][condition]) {
+            matched = true;
+            trackExportConditions(dependency, exports, sub, condition);
+            break;
+          }
         }
-      }
-    },
-    async ExportNamedDeclaration(node) {
-      const sourceValue = node && node.source ? node.source.value : '';
-
-      if (sourceValue !== '' && sourceValue.indexOf('http') !== 0) {
-        // handle relative specifier
-        if (sourceValue.indexOf('.') === 0) {
-          const entry = `/node_modules/${await resolveRelativeSpecifier(sourceValue, modulePath, dependency)}`;
 
-          updateImportMap(path.join(dependency, sourceValue), entry);
-        } else {
-          // handle bare specifier
-          updateImportMap(sourceValue, `/node_modules/${sourceValue}`);
+        if (!matched) {
+          // TODO what to do here?  what else is there besides default?
+          // console.log(`unsupported condition \`${exports[sub]}\` for dependency => \`${dependency}\``);
         }
-      }
-    },
-    async ExportAllDeclaration(node) {
-      const sourceValue = node && node.source ? node.source.value : '';
-
-      if (sourceValue !== '' && sourceValue.indexOf('http') !== 0) {
-        if (sourceValue.indexOf('.') === 0) {
-          const entry = `/node_modules/${await resolveRelativeSpecifier(sourceValue, modulePath, dependency)}`;
-
-          updateImportMap(path.join(dependency, sourceValue), entry);
+      } else {
+        // handle (unconditional) subpath exports
+        if (sub === '.') {
+          updateImportMap(dependency, `/node_modules/${dependency}/${exports[sub]}`);
+        } else if (sub.indexOf('*') >= 0) {
+          await walkExportPatterns(dependency, sub, exports[sub], resolvedRoot);
         } else {
-          updateImportMap(sourceValue, `/node_modules/${sourceValue}`);
+          updateImportMap(`${dependency}/${sub}`, `/node_modules/${dependency}/${sub}`);
         }
       }
     }
-  });
+  } else if (module || main) {
+    updateImportMap(dependency, `/node_modules/${dependency}/${module ?? main}`);
+  } else {
+    // TODO warn about no exports found
+  }
 }
 
+// https://nodejs.org/api/packages.html#package-entry-points
 async function walkPackageJson(packageJson = {}) {
-  // while walking a package.json we need to find its entry point, e.g. index.js
-  // and then walk that for import / export statements
-  // and walk its package.json for its dependencies
-
-  for (const dependency of Object.keys(packageJson.dependencies || {})) {
-    const dependencyPackageRootPath = path.join(process.cwd(), 'node_modules', dependency);
-    const dependencyPackageJsonPath = path.join(dependencyPackageRootPath, 'package.json');
-    const dependencyPackageJson = JSON.parse(fs.readFileSync(dependencyPackageJsonPath, 'utf-8'));
-    const entry = await getPackageEntryPath(dependencyPackageJson);
-    const isJavascriptPackage = Array.isArray(entry) || typeof entry === 'string' && entry.endsWith('.js') || entry.endsWith('.mjs');
-
-    if (isJavascriptPackage) {
-      const absoluteNodeModulesLocation = await getNodeModulesLocationForPackage(dependency);
-
-      // https://nodejs.org/api/packages.html#packages_determining_module_system
-      if (Array.isArray(entry)) {
-        // we have an exportMap
-        const exportMap = entry;
-
-        for (const entry of exportMap) {
-          const exportMapEntry = dependencyPackageJson.exports[entry];
-          let packageExport;
-
-          if (Array.isArray(exportMapEntry)) {
-            let fallbackPath;
-            let esmPath;
-
-            exportMapEntry.forEach((mapItem) => {
-              switch (typeof mapItem) {
-
-                case 'string':
-                  fallbackPath = mapItem;
-                  break;
-                case 'object':
-                  const entryTypes = Object.keys(mapItem);
-
-                  if (entryTypes.import) {
-                    esmPath = entryTypes.import;
-                  } else if (entryTypes.require) {
-                    console.error('The package you are importing needs commonjs support.  Please use our commonjs plugin to fix this error.');
-                    fallbackPath = entryTypes.require;
-                  } else if (entryTypes.default) {
-                    console.warn('The package you are requiring may need commonjs support.  If this module is not working for you, consider adding our commonjs plugin.');
-                    fallbackPath = entryTypes.default;
-                  }
-                  break;
-                default:
-                  console.warn(`Sorry, we were unable to detect the module type for ${mapItem} :(.  please consider opening an issue to let us know about your use case.`);
-                  break;
-
-              }
-            });
-
-            packageExport = esmPath
-              ? esmPath
-              : fallbackPath;
-          } else if (exportMapEntry.import || exportMapEntry.default) {
-            packageExport = exportMapEntry.import
-              ? exportMapEntry.import
-              : exportMapEntry.default;
-
-            // use the dependency itself as an entry in the importMap
-            if (entry === '.') {
-              updateImportMap(dependency, `/node_modules/${path.join(dependency, packageExport)}`);
-            }
-          } else if (exportMapEntry.endsWith && (exportMapEntry.endsWith('.js') || exportMapEntry.endsWith('.mjs')) && exportMapEntry.indexOf('*') < 0) {
-            // is probably a file, so _not_ an export array, package.json, or wildcard export
-            packageExport = exportMapEntry;
-          }
+  try {
+    for (const dependency of Object.keys(packageJson.dependencies || {})) {
+      const resolved = resolveBareSpecifier(dependency);
 
-          if (packageExport) {
-            const packageExportLocation = path.resolve(absoluteNodeModulesLocation, packageExport);
-
-            if (packageExport.endsWith('js')) {
-              updateImportMap(path.join(dependency, entry), `/node_modules/${path.join(dependency, packageExport)}`);
-            } else if (fs.lstatSync(packageExportLocation).isDirectory()) {
-              fs.readdirSync(packageExportLocation)
-                .filter(file => file.endsWith('.js') || file.endsWith('.mjs'))
-                .forEach((file) => {
-                  updateImportMap(path.join(dependency, packageExport, file), `/node_modules/${path.join(dependency, packageExport, file)}`);
-                });
-            } else {
-              console.warn('Warning, not able to handle export', path.join(dependency, packageExport));
-            }
-          }
-        }
+      if (resolved) {
+        const resolvedRoot = derivePackageRoot(dependency, resolved);
+        const resolvedPackageJson = (await import(new URL('./package.json', resolvedRoot), { with: { type: 'json' } })).default;
 
-        await walkPackageJson(dependencyPackageJson);
-      } else {
-        const packageEntryPointPath = path.join(absoluteNodeModulesLocation, entry);
+        walkPackageForExports(dependency, resolvedPackageJson, resolvedRoot);
 
-        // sometimes a main file is actually just an empty string... :/
-        if (fs.existsSync(packageEntryPointPath)) {
-          updateImportMap(dependency, `/node_modules/${path.join(dependency, entry)}`);
+        if (resolvedPackageJson.dependencies) {
+          for (const dependency in resolvedPackageJson.dependencies) {
+            const resolved = resolveBareSpecifier(dependency);
 
-          await walkModule(packageEntryPointPath, dependency);
-          await walkPackageJson(dependencyPackageJson);
+            if (resolved) {
+              const resolvedRoot = derivePackageRoot(dependency, resolved);
+              const resolvedPackageJson = (await import(new URL('./package.json', resolvedRoot), { with: { type: 'json' } })).default;
+
+              walkPackageForExports(dependency, resolvedPackageJson, resolvedRoot);
+
+              await walkPackageJson(resolvedPackageJson);
+            }
+          }
         }
       }
     }
+  } catch (e) {
+    console.error('Error building up import map', e);
   }
 
   return importMap;
 }
 
+// could probably go somewhere else, in a util?
 function mergeImportMap(html = '', map = {}, shouldShim = false) {
   const importMapType = shouldShim ? 'importmap-shim' : 'importmap';
   const hasImportMap = html.indexOf(`script type="${importMapType}"`) > 0;
@@ -245,7 +241,6 @@ function mergeImportMap(html = '', map = {}, shouldShim = false) {
 }
 
 export {
-  mergeImportMap,
   walkPackageJson,
-  walkModule
+  mergeImportMap
 };
>>>>>>> 7135a50 (adopt pnpm)
