diff --git a/src/index.js b/src/index.js
index 594e70bbce7ef2b1ade0e541baea052b096911b0..d9a2b2ee0e086370afe14c60153d16fe39039ac6 100755
--- a/src/index.js
+++ b/src/index.js
@@ -1,16 +1,33 @@
+function getMatchingDynamicSsrRoute(graph, pathname) {
+  return graph.find((node) => {
+    return (
+      (pathname !== "/404/") !== "/404/" &&
+      node.segment &&
+      new URLPattern({ pathname: node.segment.pathname }).test(`https://example.com${pathname}`)
+    );
+  });
+}
+
 class LitHydrationResource {
   constructor(compilation, options) {
     this.compilation = compilation;
     this.options = options;
   }
 
-  async shouldIntercept(url) {
+  async shouldIntercept(url, request, response) {
     const { pathname } = url;
+
+    if (response.headers?.get("Content-Type")?.indexOf("text/html") < 0) {
+      return false;
+    }
+
+    const matchingRouteWithSegment = getMatchingDynamicSsrRoute(this.compilation.graph, pathname);
     const matchingRoute = this.compilation.graph.find((node) => node.route === pathname);
 
     return (
-      matchingRoute &&
-      ((matchingRoute.isSSR && matchingRoute.hydration) || this.compilation.config.prerender)
+      matchingRouteWithSegment ||
+      (matchingRoute &&
+        ((matchingRoute.isSSR && matchingRoute.hydration) || this.compilation.config.prerender))
     );
   }
 
