diff --git a/node_modules/@greenwood/plugin-adapter-vercel/src/index.js b/node_modules/@greenwood/plugin-adapter-vercel/src/index.js
index f37e815..e498b8a 100644
--- a/node_modules/@greenwood/plugin-adapter-vercel/src/index.js
+++ b/node_modules/@greenwood/plugin-adapter-vercel/src/index.js
@@ -87,7 +87,7 @@ async function vercelAdapter(compilation) {
     const { id } = page;
     const outputRoot = new URL(`./${basePath}/${id}.func/`, adapterOutputUrl);
     const files = (await fs.readdir(outputDir))
-      .filter(file => file.startsWith(`${id}.route.chunk.`) && file.endsWith('.js'));
+      .filter(file => file.startsWith(`${id}.route.`) && file.endsWith('.js'));
 
     await setupFunctionBuildFolder(id, outputType, outputRoot);
 
