diff --git a/src/index.js b/src/index.js
index 4efca4b85ee5947683a316b529c5863a46e28a9b..8bc2da7f6f5751c480c58d228479cbd615014d04 100755
--- a/src/index.js
+++ b/src/index.js
@@ -7,9 +7,9 @@ class LitHydrationResource extends ResourceInterface {
 
   async shouldIntercept(url) {
     const { pathname } = url;
-    const matchingRoute = this.compilation.graph.find((node) => node.route === pathname) || {};
-
-    return matchingRoute.isSSR && matchingRoute.hydration;
+    const matchingRoute = this.compilation.graph.find((node) => node.route === pathname);
+  
+    return matchingRoute && ((matchingRoute.isSSR && matchingRoute.hydration) || this.compilation.config.prerender);
   }
 
   async intercept(url, request, response) {
