diff --git a/src/types/index.d.ts b/src/types/index.d.ts
index 5f5384ea145e4a41b126492b586fc77c598bec62..65587930f78db78fa1ade57a28aa52d7620c604c 100644
--- a/src/types/index.d.ts
+++ b/src/types/index.d.ts
@@ -1,4 +1,4 @@
-import type { Plugin } from "@greenwood/cli";
+import type { AdapterPlugin } from "@greenwood/cli";
 
 type SUPPORTED_NODE_VERSIONS = "nodejs22.x" | "nodejs20.x" | "nodejs18.x";
 
@@ -6,7 +6,7 @@ type ImportRawPluginOptions = {
   matches?: string[];
 };
 
-export type ImportRawPlugin = (options?: ImportRawPluginOptions) => Array<Plugin>;
+export type ImportRawPlugin = (options?: ImportRawPluginOptions) => AdapterPlugin;
 
 declare module "@greenwood/plugin-import-raw" {
   export const greenwoodPluginImportRaw: ImportRawPlugin;
