diff --git a/src/execute-route-module.js b/src/execute-route-module.js
index 45ce49efc5e2e0133c59db539e334aad1b982739..29870d5663e4a072f544e95d12e099858670bbbc 100644
--- a/src/execute-route-module.js
+++ b/src/execute-route-module.js
@@ -1,8 +1,8 @@
-import { render, LitElementRenderer } from "@lit-labs/ssr";
+import "@lit-labs/ssr-dom-shim/register-css-hook.js";
+const { render, LitElementRenderer } = await import('@lit-labs/ssr');
 import { collectResult } from "@lit-labs/ssr/lib/render-result.js";
 import { html, literal, unsafeStatic } from "lit/static-html.js";
 import { unsafeHTML } from "lit/directives/unsafe-html.js";
-import "@lit-labs/ssr-dom-shim/register-css-hook.js";
 
 async function executeRouteModule({
   moduleUrl,
