diff --git a/src/api-route-worker.js b/src/api-route-worker.js
new file mode 100644
index 0000000000000000000000000000000000000000..d49b9efa056b7361856d29dfc13a5bfb6e4f0d2d
--- /dev/null
+++ b/src/api-route-worker.js
@@ -0,0 +1,62 @@
+// https://github.com/nodejs/modules/issues/307#issuecomment-858729422
+import { parentPort } from "node:worker_threads";
+import { transformKoaRequestIntoStandardRequest } from "@greenwood/cli/src/lib/resource-utils.js";
+import "@lit-labs/ssr-dom-shim/register-css-hook.js";
+
+// based on https://stackoverflow.com/questions/57447685/how-can-i-convert-a-request-object-into-a-stringifiable-object-in-javascript
+// TODO: make these a shared utility
+async function responseAsObject(response) {
+  if (!(response instanceof Response)) {
+    throw Object.assign(new Error(), {
+      name: "TypeError",
+      message: "Argument must be a Response object",
+    });
+  }
+  response = response.clone();
+
+  function stringifiableObject(obj) {
+    const filtered = {};
+    for (const key in obj) {
+      if (["boolean", "number", "string"].includes(typeof obj[key]) || obj[key] === null) {
+        filtered[key] = obj[key];
+      }
+    }
+    return filtered;
+  }
+
+  return {
+    ...stringifiableObject(response),
+    headers: Object.fromEntries(response.headers),
+    body: await response.text(),
+  };
+}
+
+async function executeRouteModule({ href, request, params }) {
+  const { body, headers = {}, method, url } = request;
+  const contentType = headers["content-type"] || "";
+  // @ts-expect-error see https://github.com/microsoft/TypeScript/issues/42866
+  const { handler } = await import(new URL(href));
+  const format = contentType.startsWith("application/json") ? JSON.parse(body) : body;
+
+  // handling of serialized FormData across Worker threads
+  if (contentType.startsWith("x-greenwood/www-form-urlencoded")) {
+    headers["content-type"] = "application/x-www-form-urlencoded";
+  }
+
+  const response = await handler(
+    transformKoaRequestIntoStandardRequest(new URL(url), {
+      method,
+      header: headers,
+      body: format,
+    }),
+    {
+      params,
+    },
+  );
+
+  parentPort.postMessage(await responseAsObject(response));
+}
+
+parentPort.on("message", async (task) => {
+  await executeRouteModule(task);
+});
diff --git a/src/index.js b/src/index.js
index 594e70bbce7ef2b1ade0e541baea052b096911b0..bb586af7c32f7f33b4f4a0d6c836da7f906e5260 100755
--- a/src/index.js
+++ b/src/index.js
@@ -76,6 +76,7 @@ const greenwoodPluginRendererLit = () => {
       provider: () => {
         return {
           executeModuleUrl: new URL("./execute-route-module.js", import.meta.url),
+          apiRouteWorkerUrl: new URL("./api-route-worker.js", import.meta.url),
         };
       },
     },
