diff --git a/src/lib/layout-utils.js b/src/lib/layout-utils.js
index 40b9d585769c4789efd3ff4a24bc90f483e15ac9..9cb01bf6c25016e9b750c06c472aed26af59def3 100644
--- a/src/lib/layout-utils.js
+++ b/src/lib/layout-utils.js
@@ -47,8 +47,8 @@ async function mergeContentIntoLayout(
   matchingRoute,
 ) {
   const activeFrontmatterTitleKey = "${globalThis.page.title}";
-  const parentRoot = parentContents && parse(parentContents);
-  const childRoot = parse(childContents);
+  const parentRoot = parentContents && parse(parentContents, { comment: true });
+  const childRoot = parse(childContents, { comment: true });
   let mergedContents = "";
 
   if ((parentContents && !valid(parentContents)) || (childContents && !valid(childContents))) {
