diff --git a/src/types/index.d.ts b/src/types/index.d.ts
index abffd41fec3fb2b8324b5058c2b336aac2cf0f25..e21f89aa554345b337d15925f0192095466a1931 100644
--- a/src/types/index.d.ts
+++ b/src/types/index.d.ts
@@ -1,4 +1,4 @@
-import type { Plugin } from "@greenwood/cli";
+import type { AdapterPlugin } from "@greenwood/cli";
 
 type SUPPORTED_NODE_VERSIONS = "nodejs22.x" | "nodejs20.x" | "nodejs18.x";
 
@@ -6,7 +6,7 @@ type VercelAdapterOptions = {
   runtime?: SUPPORTED_NODE_VERSIONS;
 };
 
-export type VercelAdapter = (options?: VercelAdapterOptions) => Array<Plugin>;
+export type VercelAdapter = (options?: VercelAdapterOptions) => AdapterPlugin;
 
 declare module "@greenwood/plugin-adapter-vercel" {
   export const greenwoodPluginAdapterVercel: VercelAdapter;
