diff --git a/node_modules/@greenwood/cli/src/config/rollup.config.js b/node_modules/@greenwood/cli/src/config/rollup.config.js
index b0ea175..e9103ed 100644
--- a/node_modules/@greenwood/cli/src/config/rollup.config.js
+++ b/node_modules/@greenwood/cli/src/config/rollup.config.js
@@ -400,8 +400,8 @@ const getRollupConfigForSsr = async (compilation, input) => {
     input: filepath,
     output: {
       dir: normalizePathnameForWindows(outputDir),
-      entryFileNames: '[name].route.js',
-      chunkFileNames: '[name].route.chunk.[hash].js'
+      entryFileNames: `${path.basename(filepath).split('.')[0]}.route.js`,
+      chunkFileNames: `${path.basename(filepath).split('.')[0]}.route.chunk.[hash].js`
     },
     plugins: [
       greenwoodResourceLoader(compilation),
