diff --git a/src/lifecycles/graph.js b/src/lifecycles/graph.js
index bdb495048d6ed785a433ca92bad407ffde0bf9d6..56754871fd4265c0b5d59abd219f45d8719fdf24 100644
--- a/src/lifecycles/graph.js
+++ b/src/lifecycles/graph.js
@@ -109,7 +109,9 @@ const generateGraph = async (compilation) => {
           }
 
           // TODO should API routes be run in isolation mode like SSR pages?
-          const { isolation } = await import(filenameUrl).then((module) => module);
+          // const { isolation } = await import(filenameUrl).then((module) => module);
+          const contents = await fs.readFile(filenameUrl, "utf8");
+          const isolation = contents.indexOf('export const isolation = true;') >= 0;
           const { segmentKey, dynamicRoute } = getDynamicSegmentsFromRoute({
             route,
             relativePagePath,
