diff --git a/node_modules/@greenwood/cli/src/config/rollup.config.js b/node_modules/@greenwood/cli/src/config/rollup.config.js
index df59ef6..8653314 100644
--- a/node_modules/@greenwood/cli/src/config/rollup.config.js
+++ b/node_modules/@greenwood/cli/src/config/rollup.config.js
@@ -138,6 +138,7 @@ function greenwoodPatchSsrPagesEntryPointRuntimeImport() {
     name: 'greenwood-patch-ssr-pages-entry-point-runtime-import',
     generateBundle(options, bundle) {
       Object.keys(bundle).forEach((key) => {
+        console.log({ key });
         if (key.startsWith('__')) {
           // ___GWD_ENTRY_FILE_URL=${filename}___
           const needle = bundle[key].code.match(/___GWD_ENTRY_FILE_URL=(.*.)___/);
@@ -146,7 +147,7 @@ function greenwoodPatchSsrPagesEntryPointRuntimeImport() {
 
             bundle[key].code = bundle[key].code.replace(/'___GWD_ENTRY_FILE_URL=(.*.)___'/, `new URL('./_${entryPathMatch}', import.meta.url)`);
           } else {
-            console.warn(`Could not find entry path match for bundle => ${ley}`);
+            console.warn(`Could not find entry path match for bundle => ${key}`);
           }
         }
       });
